### PR TITLE
[nis] feat: allow fork height override

### PR DESCRIPTION
**Pull request by Wayonb**
problem: new testnet does not work with NEM wallet
solution: allow the fork heights to be override from the nis configuration.
               setting the fork height to 1 for all forks will create a testnet
                with the same functionally as mainnet

This is the initial PR for the fork changes.  
1. extend the current ForkConfiguration class to include all fork info
2. plump the ForkConfiguration class the all the fork location
3. default fork height is left unchanged.
4. fork height can be overridden in the nis configuration file
 
Ran a test with the NEM Wallet and was able to submit a transfer tx at block 3 when all the fork height is set to 1.  Feedback welcome but still need to run some more tests.
Screenshot from 2023-03-10 01-10-57

### CODE REVIEW

**Comment by Wayonb** [comment]
testing looking fine.
1. sync with testnet
2. currently syncing with mainnet
Screenshot from 2023-03-14 00-04-51

File: nis/src/test/java/org/nem/nis/ForkConfigurationTest.java [hunk 1]
```diff
@@ -60,22 +77,42 @@ private static void cannotParseWithInvalidHashes(final String separator, final S
...
 
 		// Act + Assert
-		ExceptionAssert.assertThrows(v -> new ForkConfiguration(new NemProperties(properties)), CryptoException.class);
+		ExceptionAssert.assertThrows(v -> new ForkConfiguration.Builder(new NemProperties(properties)).build(), CryptoException.class);
 	}
 
-	// endregion
-
-	// region basic
+	private static void canReadConfiguration(final int version, final Supplier<ForkConfiguration> configSupplier) {
```
Comment (Jaguar0625) [inline nis/src/test/java/org/nem/nis/ForkConfigurationTest.java:83]: rename assertCan...

File: nis/src/test/java/org/nem/nis/ForkConfigurationTest.java [hunk 2]
```diff
@@ -60,22 +77,42 @@ private static void cannotParseWithInvalidHashes(final String separator, final S
...
+	// endregion
+
+	// region basic
+
+	@Test
+	public void canCreateDefaultConfiguration() {
+		// Arrange:
+		final int version = NetworkInfos.getDefault().getVersion() << 24;
+
+		canReadConfiguration(version, () -> new ForkConfiguration.Builder().build());;
```
Comment (Jaguar0625) [inline nis/src/test/java/org/nem/nis/ForkConfigurationTest.java:115]: // Act +Assert:

File: nis/src/test/java/org/nem/nis/ForkConfigurationTest.java [hunk 3]
```diff
@@ -92,26 +129,40 @@ public void canCreateCustomConfiguration() {
...
+				.secondFeeForkHeight(new BlockHeight(6)).build();
 
 		// Assert:
 		MatcherAssert.assertThat(config.getTreasuryReissuanceForkHeight(), IsEqual.equalTo(new BlockHeight(1234)));
 		MatcherAssert.assertThat(config.getTreasuryReissuanceForkTransactionHashes(), IsEqual.equalTo(hashes));
 		MatcherAssert.assertThat(config.getTreasuryReissuanceForkFallbackTransactionHashes(), IsEqual.equalTo(fallbackHashes));
+
+		MatcherAssert.assertThat(config.getFeeForkHeight(), IsEqual.equalTo(new BlockHeight(1)));
+		MatcherAssert.assertThat(config.getMosaicsForkHeight(), IsEqual.equalTo(new BlockHeight(2)));
+		MatcherAssert.assertThat(config.getmultisigMOfNForkHeight(), IsEqual.equalTo(new BlockHeight(3)));
```
Comment (Jaguar0625) [inline nis/src/test/java/org/nem/nis/ForkConfigurationTest.java:145]: getMultisig...

File: nis/src/test/java/org/nem/nis/ForkConfigurationTest.java [hunk 4]
```diff
@@ -92,26 +129,40 @@ public void canCreateCustomConfiguration() {
...
 
 	@Test
 	public void canReadDefaultConfiguration() {
 		// Arrange:
-		final Properties properties = new Properties();
+		final int version = NetworkInfos.getDefault().getVersion() << 24;
 
-		// Act:
-		final ForkConfiguration config = new ForkConfiguration(new NemProperties(properties));
+		canReadConfiguration(version, () -> new ForkConfiguration.Builder(new NemProperties(new Properties())).build());;
```
Comment (Jaguar0625) [inline nis/src/test/java/org/nem/nis/ForkConfigurationTest.java:156]: double semicolon (few places in file)

File: nis/src/test/java/org/nem/nis/ForkConfigurationTest.java [hunk 5]
```diff
@@ -92,26 +129,40 @@ public void canCreateCustomConfiguration() {
...
-		MatcherAssert.assertThat(config.getTreasuryReissuanceForkHeight(), IsEqual.equalTo(new BlockHeight(1)));
-		MatcherAssert.assertThat(config.getTreasuryReissuanceForkTransactionHashes(), IsEqual.equalTo(new ArrayList<Hash>()));
-		MatcherAssert.assertThat(config.getTreasuryReissuanceForkFallbackTransactionHashes(), IsEqual.equalTo(new ArrayList<Hash>()));
+	@Test
+	public void canReadMainnetConfiguration() {
+		// Arrange:
+		final NetworkInfo networkInfo = NetworkInfos.getMainNetworkInfo();
+		final int version = networkInfo.getVersion() << 24;
+
+		canReadConfiguration(version, () -> new ForkConfiguration.Builder(new NemProperties(new Properties()), networkInfo).build());
```
Comment (Jaguar0625) [inline nis/src/test/java/org/nem/nis/ForkConfigurationTest.java:165]: should there be separate ctor that accepts only networkInfo ?

when would well-defined network properties need to be overridden?

File: nis/src/test/java/org/nem/nis/cache/NamespaceCacheTest.java [hunk 6]
```diff
@@ -970,13 +971,14 @@ private void assertBasicCopy(final Function<T, T> copyCache) {
 	public void getReturnsExpectedNemNamespaceEntry() {
 		// Arrange:
 		final T cache = this.createCache();
+		final BlockHeight mosaicRedefinitionForkHeight = new ForkConfiguration.Builder().build().getMosaicRedefinitionForkHeight();
```
Comment (Jaguar0625) [inline nis/src/test/java/org/nem/nis/cache/NamespaceCacheTest.java:974]: unused?

File: nis/src/main/java/org/nem/nis/NemNamespaceEntry.java [hunk 7]
```diff
@@ -0,0 +1,88 @@
...
+
+	/**
+	 * Sets the namespace entry for 'nem' that contains a single mosaic 'nem.xem'.
+	 *
+	 * @param mosaicRedefinitionForkHeight The mosaic redefinition fork height.
+	 */
+	public static void setDefault(final BlockHeight mosaicRedefinitionForkHeight) {
+		instance.set(createNemNamespaceEntry(mosaicRedefinitionForkHeight));
+	}
+}
```
Comment (Jaguar0625) [inline nis/src/main/java/org/nem/nis/NemNamespaceEntry.java:88]: should you add tests for get/setDefault (see NetworkInfosTest.java)

**Review by Jaguar0625 — CHANGES_REQUESTED** [review]

File: nis/src/test/java/org/nem/nis/ForkConfigurationTest.java (hunk 5 above)
Comment (Wayonb) [reply nis/src/test/java/org/nem/nis/ForkConfigurationTest.java:165]: currently the network is only overridden from the nis configuration.  So the Nem properties are also available.

Yea, can create a ctor that only accepts networkinfo

File: nis/src/main/java/org/nem/nis/NemNamespaceEntry.java (hunk 7 above)
Comment (Wayonb) [reply nis/src/main/java/org/nem/nis/NemNamespaceEntry.java:88]: yea, will add some

File: nis/src/main/java/org/nem/nis/FeeForkDto.java [hunk 8]
```diff
@@ -0,0 +1,40 @@
+package org.nem.nis;
+
+import org.nem.core.model.primitive.BlockHeight;
+
+/**
+ * DTO for the fee fork heights.
+ */
+public class FeeForkDto {
```
Comment (Jaguar0625) [inline nis/src/main/java/org/nem/nis/FeeForkDto.java:8]: not sure if we need Dto in the name vs just FeeFork(s) 

cc: @gimre-xymcity ?

File: nis/src/main/java/org/nem/nis/FeeForkDto.java [hunk 9]
```diff
@@ -0,0 +1,40 @@
...
+	public final BlockHeight firstHeight;
+	public final BlockHeight secondHeight;
+
+	/**
+	 * Creates a fee fork heights.
```
Comment (Jaguar0625) [inline nis/src/main/java/org/nem/nis/FeeForkDto.java:13]: ... object / dto.

File: nis/src/main/java/org/nem/nis/FeeForkDto.java [hunk 10]
```diff
@@ -0,0 +1,40 @@
...
+
+	/**
+	 * Gets the second fee fork height.
+	 *
+	 * @return The second fee fork height.
+	 */
+	public BlockHeight getSecondHeight() {
+		return this.secondHeight;
+	}
+}
```
Comment (Jaguar0625) [inline nis/src/main/java/org/nem/nis/FeeForkDto.java:40]: could add simple test for this (ctor), if desired

File: nis/src/main/java/org/nem/nis/ForkConfiguration.java [hunk 11]
```diff
@@ -19,21 +19,19 @@ public class ForkConfiguration {
 
 	private final BlockHeight multisigMOfNForkHeight;
 	private final BlockHeight mosaicsForkHeight;
-	private final BlockHeight feeForkHeight;
 	private final BlockHeight remoteAccountForkHeight;
 	private final BlockHeight mosaicRedefinitionForkHeight;
-	private final BlockHeight secondFeeForkHeight;
+	private final FeeForkDto feeForkDto;
```
Comment (Jaguar0625) [inline nis/src/main/java/org/nem/nis/ForkConfiguration.java:24]: definitely would drop dto postfix from member / variable / arg names

cc: @gimre-xymcity

File: nis/src/test/java/org/nem/nis/ForkConfigurationTest.java [hunk 12]
```diff
@@ -60,22 +77,44 @@ private static void cannotParseWithInvalidHashes(final String separator, final S
...
+
+	// region basic
+
+	@Test
+	public void canCreateDefaultConfiguration() {
+		// Arrange:
+		final int version = NetworkInfos.getDefault().getVersion() << 24;
+
+		// Act +Assert:
+		assertCanReadConfiguration(version, () -> new ForkConfiguration.Builder().build());;
```
Comment (Jaguar0625) [inline nis/src/test/java/org/nem/nis/ForkConfigurationTest.java:117]: double semicolon

**Review by Jaguar0625 — APPROVED** [review]

**Pull request merged**